### PR TITLE
clean: Dockerfile using multistage build

### PR DESCRIPTION
**Pull request by josemiguelmelo**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,26 +3,20 @@ ARG NODE_IMAGE_VERSION=13.5.0-alpine
...
 
 COPY --from=builder dist dist
-COPY package*.json ./
-COPY docs /docs
-
-RUN npm install --production
-
-COPY docs /docs
+COPY --from=builder node_modules node_modules
+COPY --from=builder package*.json ./
```
Comment (peavers) [inline Dockerfile]: I assume the wildcard is here for the lock file? Do we really need to copy the lock file? If so, is the wildcard really needed for it?

Comment (josemiguelmelo) [reply Dockerfile]: Yes, the wildcard will include the lock file.
We need it to ensure the dependency tree is the same identical tree we use, regardless of intermediate dependency updates.

Comment (lolgab) [reply Dockerfile]: You can explicitly copy package-lock.json and package.json in two steps

File: package.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
   "scripts": {
     "compile": "tsc",
     "generateDocs": "npm run compile && node dist/src/docGeneratorMain.js",
+    "start": "npm run compile && node dist/src/index.js",
```
Comment (lolgab) [inline package.json]: This is useless since the tool requires thing from the environment (`/src` directory etc.)

File: Dockerfile [hunk 3]
```diff
@@ -3,26 +3,20 @@ ARG NODE_IMAGE_VERSION=13.5.0-alpine
...
 RUN npm test
+RUN npm run generateDocs
 
 FROM node:$NODE_IMAGE_VERSION
 
 COPY --from=builder dist dist
-COPY package*.json ./
-COPY docs /docs
-
-RUN npm install --production
```
Comment (lolgab) [inline Dockerfile]: Is this gone?

File: Dockerfile (hunk 1 above)
Comment (josemiguelmelo) [reply Dockerfile]: Done ✅

File: package.json (hunk 2 above)
Comment (josemiguelmelo) [reply package.json]: Removed it

File: Dockerfile [hunk 4]
```diff
@@ -3,26 +3,21 @@ ARG NODE_IMAGE_VERSION=13.5.0-alpine
...
-
-COPY docs /docs
+COPY --from=builder node_modules node_modules
```
Comment (lolgab) [inline Dockerfile]: If you do this it's useless to have multiple stages.
And you are copying even the devDependencies

Comment (josemiguelmelo) [reply Dockerfile]: You're right. Added the `npm install --production` and removed this COPY.

File: Dockerfile (hunk 3 above)
Comment (josemiguelmelo) [reply Dockerfile]: Added it ✅

File: Dockerfile [hunk 5]
```diff
@@ -3,27 +3,23 @@ ARG NODE_IMAGE_VERSION=13.5.0-alpine
...
+RUN npm run generateDocs
 
 FROM node:$NODE_IMAGE_VERSION
 
 COPY --from=builder dist dist
-COPY package*.json ./
-COPY docs /docs
+COPY --from=builder package.json ./
+COPY --from=builder package-lock.json ./
+COPY --from=builder docs docs
```
Comment (lolgab) [inline Dockerfile]: This is good, since it uses the newly generated docs!

**Pull request merged**